### PR TITLE
Adding reconcile logic for TLS and CA certs for App & Console

### PR DESCRIPTION
**Pull request by vishnuchalla**
## Description

Adding logic to reconcile for TLS and CA certs changes across the system.

## Type of change

- [x] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Closes # https://issues.redhat.com/browse/OLS-424

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [x] If it is a core feature, I have added thorough tests.

## Testing
Tested in local by recreating and editing the secrets. Also made sure unit tests cover all the cases too.

### CODE REVIEW

**Comment by vishnuchalla** [comment]
@raptorsun @vbelouso @xrajesh Do we need any documentation for this? Just trying to get some thoughts.

**Comment by raptorsun** [comment]
> @raptorsun @vbelouso @xrajesh Do we need any documentation for this? Just trying to get some thoughts.

It is great to have some explaination, maybe put into file `Readme.md`?

File: internal/controller/ols_app_server_reconciliator.go [hunk 1]
```diff
@@ -240,3 +245,34 @@ func (r *OLSConfigReconciler) reconcileService(ctx context.Context, cr *olsv1alp
 	r.logger.Info("OLS service reconciled", "service", service.Name)
 	return nil
 }
+
+func (r *OLSConfigReconciler) reconcileTLSSecret(ctx context.Context, cr *olsv1alpha1.OLSConfig) error {
+	var secretValues map[string]string
+	foundSecret := &corev1.Secret{}
+	err := r.Client.Get(ctx, client.ObjectKey{Name: OLSCertsSecretName, Namespace: r.Options.Namespace}, foundSecret)
+	if err != nil {
+		return fmt.Errorf("failed to fetch secret: %w", err)
```
Comment (xrajesh) [inline internal/controller/ols_app_server_reconciliator.go:254]: @vishnuchalla - Can we update this error message - This is caused most likely due the tls-secret not present.

File: internal/controller/ols_app_server_reconciliator.go [hunk 2]
```diff
@@ -240,3 +245,34 @@ func (r *OLSConfigReconciler) reconcileService(ctx context.Context, cr *olsv1alp
...
+	}
+	secretValues, err = getSecretContent(r.Client, OLSCertsSecretName, r.Options.Namespace, []string{"tls.key", "tls.crt"})
+	if err != nil {
+		return fmt.Errorf("failed fetching tls certs from secret: %s. error: %w", OLSCertsSecretName, err)
```
Comment (xrajesh) [inline internal/controller/ols_app_server_reconciliator.go:258]: @vishnuchalla Here too . If we can give a updated message - `failed` seem to lead that its a operator failure. Thinking - `Secret does not have expected tls.key or tls.crt` ?

Comment (vishnuchalla) [reply internal/controller/ols_app_server_reconciliator.go]: Sounds good. Updated

**Comment by raptorsun** [comment]
need rebase on main branch, removing redis related codes.

File: internal/controller/constants.go [hunk 3]
```diff
@@ -45,6 +45,16 @@ const (
...
+	// OLSAppTLSHashKey is the key of the hash value of the OLS App TLS certificates
+	OLSAppTLSHashKey = "hash/olstls"
+	// OLSCAHashKey is the key of the hash value of the OLS CA certificate
+	OLSCAHashKey = "hash/ca"
+	// OLSConsoleTLSHashKey is the key of the hash value of the OLS Console TLS certificates
+	OLSConsoleTLSHashKey = "hash/olsconsoletls"
+	// RedisTLSHashKey is the key of the hash value of Redis
+	RedisTLSHashKey = "hash/redistls"
+	// RedisCAHashKey is the key of the hash value of Redis CA Cert
+	RedisCAHashKey = "hash/redisca"
```
Comment (raptorsun) [inline internal/controller/constants.go:57]: we can remove `RedisTLSHashKey` and `RedisCAHashKey` for they are no longer needed.

File: internal/controller/constants.go [hunk 4]
```diff
@@ -79,9 +89,14 @@ const (
...
-	OLSConfigHashStateCacheKey   = "olsconfigmap-hash"
-	LLMProviderHashStateCacheKey = "llmprovider-hash"
-	RedisConfigHashStateCacheKey = "olsredisconfig-hash"
+	OLSAppTLSHashStateCacheKey     = "olsapptls-hash"
+	OLSConfigHashStateCacheKey     = "olsconfigmap-hash"
+	OLSConsoleTLSHashStateCacheKey = "olsconsoletls-hash"
+	OLSCAHashStateCacheKey         = "olsca-hash"
+	RedisConfigHashStateCacheKey   = "olsredisconfig-hash"
+	RedisTLSHashStateCacheKey      = "redistls-hash"
+	RedisCAHashStateCacheKey       = "redisca-hash"
```
Comment (raptorsun) [inline internal/controller/constants.go:98]: we do not need them, either.

File: internal/controller/ols_app_server_reconciliator_test.go [hunk 5]
```diff
@@ -177,19 +274,17 @@ var _ = Describe("App server reconciliator", Ordered, func() {
...
 			Expect(secretCreationErr).NotTo(HaveOccurred())
+
 			olsConfig := &olsv1alpha1.OLSConfig{}
 			err = k8sClient.Get(ctx, crNamespacedName, olsConfig)
 			Expect(err).NotTo(HaveOccurred())
 
-			By("Reconcile for LLM Provider Secrets Again")
-			err = reconciler.reconcileLLMSecrets(ctx, olsConfig)
-			Expect(err).NotTo(HaveOccurred())
-
```
Comment (raptorsun) [inline internal/controller/ols_app_server_reconciliator_test.go:192]: it seems a mistake removal.

File: internal/controller/constants.go (hunk 3 above)
Comment (vishnuchalla) [reply internal/controller/constants.go:57]: The idea was to keep them as is and use them for our new cache. If we want them to be removed, lets take them in a separate PR as we have many other things to remove regarding redis in our current mainstream. Is that okay?

File: internal/controller/constants.go (hunk 4 above)
Comment (vishnuchalla) [reply internal/controller/constants.go:98]: Same as the above comment.

File: internal/controller/constants.go (hunk 3 above)
Comment (raptorsun) [reply internal/controller/constants.go:57]: ok we leave it there

**Comment by raptorsun** [comment]
need rebase again 🥲

File: internal/controller/ols_app_server_deployment.go [hunk 6]
```diff
@@ -247,17 +247,20 @@ func (r *OLSConfigReconciler) updateOLSDeployment(ctx context.Context, existingD
 
 	// Validate deployment annotations.
 	if existingDeployment.Annotations == nil ||
-		existingDeployment.Annotations[OLSConfigHashKey] != r.stateCache[OLSConfigHashStateCacheKey] ||
-		existingDeployment.Annotations[LLMProviderHashKey] != r.stateCache[LLMProviderHashStateCacheKey] {
+		existingDeployment.Annotations[OLSConfigHashKey] != r.stateCache[OLSConfigHashStateCacheKey] || existingDeployment.Annotations[OLSAppTLSHashKey] != r.stateCache[OLSAppTLSHashStateCacheKey] || existingDeployment.Annotations[OLSCAHashKey] != r.stateCache[OLSCAHashStateCacheKey] || existingDeployment.Annotations[LLMProviderHashKey] != r.stateCache[LLMProviderHashStateCacheKey] {
```
Comment (xrajesh) [inline internal/controller/ols_app_server_deployment.go:250]: @vishnuchalla - The CA Cert Hash `r.stateCache[OLSCAHashStateCacheKey]` I dont see it , getting a hash setup other than the initialization. Do we need this check ?

**Comment by xrajesh** [comment]
@vishnuchalla When the real cert rotation happens the annotation we added may be removed - will we reconcile at that time too? Thinking how that case will work.

**Comment by raptorsun** [comment]
Comments are addressed in this PR https://github.com/openshift/lightspeed-operator/pull/129
CA secret is not taken care as we are not using it in the deployment.

**Pull request closed without merging**